### PR TITLE
Feature/4105 sharing detection

### PR DESCRIPTION
**Pull request by aerych**
This PR improves how support for Sharing (via Publicize and share buttons) is detected on wpcom and Jetpack connected blogs.  The sharing feature is always on for wpcom blogs, but is either on or off for Jetpack connected blogs depending on which modules are enabled.  The testing scenarios should cover most those situations.

Some work was done to make sure options were always up to date before viewing a blog's details screen. I'm a bit iffy on `BlogSyncState` as a way to get a completion handler for `syncBlog:`, but it seemed a bit cleaner than breakout out the options sync, or implementing multiple call backs. Very open to other ideas.

Needs review: @koke 

### Testing
Scenario 1:
Sign into with a wpcom account and view the blog details screen of a wpcom blog.  Sharing should be present.

Scenario 2: 
Sign into a wpcom account that has a Jetpack connected blog. 
View that the details screen for the Jetpack connected blog.  Sharing should be present.

Scenario 3:
Be signed out of wpcom
Sign into a self-hosted blog and view its blog details screen. Sharing should *not* be present

Scenario 4:
Repeat the steps in Scenario 3.
Tap on the Stats feature. Sign in to Jetpack when prompted.
Return to the details screen.  Sharing should be present. 

Scenario 5: 
Repeat the steps in Scenario 3.
Go to the Me tab and sign into the wpcom account that is connected to the self-hosted site's Jetpack. 
Return to the My Sites tab and view the detail for the self-hosted blog. Sharing should be present. 

Scenario 6: 
Be signed into wpcom. 
Sign into a Jetpack connected blog that is connected to a *different* wpcom account. 
Go to that blog's details screen. Sharing should *not* be present. 
Tap on Stats and connect to Jetpack when prompted. 
Return to the details screen.  Sharing should be present. 

Scenario 7: 
Disable both Publicize and Sharing buttons on a Jetpack connected blog. 
Repeat the steps in Scenario 2.  Sharing should *not* be present.

Scenario 8: 
Disable either Publicize or Sharing buttons on a Jetpack connected blog. 
Repeat the steps in Scenario 2.  Sharing should be present.  Note that the Sharing VC does not yet reflect one of the features being enabled and one not. This will be handled in a future PR. 

Scenario 9: 
Disable the JSON API module on a Jetpack connected blog. 
Repeat the steps in Scenario 2.  Sharing should *not* be present.

### CODE REVIEW

File: WordPress/Classes/Models/Blog.m [hunk 1]
```diff
@@ -439,6 +444,46 @@ - (BOOL)supports:(BlogFeature)feature
     }
 }
 
+- (BOOL)supportsPublicize
+{
+    // wpcom blog and publicize has not been disabled
+    if (self.isHostedAtWPcom && ![[self getOptionValue:OptionsKeyPublicizeDisabled] boolValue]) {
+        return YES;
+    }
```
Comment (koke) [inline WordPress/Classes/Models/Blog.m]: I think if `isHostedAtWPcom` is true but publicize is disabled, it should return `NO` immediately.

File: WordPress/Classes/Models/Blog.m [hunk 2]
```diff
@@ -439,6 +444,46 @@ - (BOOL)supports:(BlogFeature)feature
...
+
+    // self-hosted connected to jetpack via the wpcom acct signed into the app.
+    if (self.dotComID && [self jetpackPublicizeModuleEnabled]) {
+        return YES;
+    }
```
Comment (koke) [inline WordPress/Classes/Models/Blog.m]: I think this `if` block is unnecessary. `jetpackRESTSupported` below already handles this case

File: WordPress/Classes/Models/Blog.m [hunk 3]
```diff
@@ -439,6 +444,46 @@ - (BOOL)supports:(BlogFeature)feature
...
+{
+    // wpcom blog
+    if (self.isHostedAtWPcom) {
+        return YES;
+    }
+
+    // self-hosted connected to jetpack via the wpcom acct signed into the app.
+    if (self.dotComID && [self jetpackSharingButtonsModuleEnabled]) {
+        return YES;
+    }
```
Comment (koke) [inline WordPress/Classes/Models/Blog.m]: Also unnecessary, like the one in `supportsPublicize`

**Comment by koke** [comment]
> I'm a bit iffy on BlogSyncState

Agreed, it smells like trouble ahed :wink:
Some issues I can detect:
- If you add/remove requests but forget to update the request count :boom: 
- Not thread safe
- We aggregate all the errors, but then nothing uses them

Some other ideas:
- Batch requests: they are problematic on self hosted, so no :stuck_out_tongue: 
- NSOperation dependencies: get an operation for each request, then have a syncBlog operation that depends on the previous 4. See `-[CreateAccountAndBlogViewController createUserAndSite]`. Requires a bit of refactor, but the best solution IMO.
- Use [dispatch_group](http://commandshift.co.uk/blog/2014/03/19/using-dispatch-groups-to-wait-for-multiple-web-services/). Doesn't require changes on the remotes or the categories service.
- If you prefer BlogSyncState, use the [OSAtomic](https://developer.apple.com/library/mac/documentation/System/Reference/OSAtomic_header_reference/#//apple_ref/c/func/OSAtomicDecrement64Barrier) functions to check and modify the counter.

File: WordPress/Classes/Networking/RemoteBlogOptionsHelper.m [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+        for (NSString *key in optionsDirectMapKeys) {
+            NSString *sourceKeyPath = [NSString stringWithFormat:@"options.%@", key];
+            if ([response valueForKeyPath:sourceKeyPath] != nil) {
+                options[key] = [response valueForKeyPath:sourceKeyPath];
+            }
+        }
+    } else {
+        //valid default values
+        options[@"software_version"] = @"3.6";
```
Comment (koke) [inline WordPress/Classes/Networking/RemoteBlogOptionsHelper.m]: I can almost see the puzzled look on some poor user's face when the app asks them to upgrade the WordPress version in WordPress.com. I wonder if `response[@"options"] == nil` should be treated as an error

File: WordPress/Classes/Models/Blog.m (hunk 1 above)
Comment (aerych) [inline WordPress/Classes/Models/Blog.m]: :+1:

File: WordPress/Classes/Models/Blog.m (hunk 2 above)
Comment (aerych) [inline WordPress/Classes/Models/Blog.m]: `jetpackRESTSupported` returns `NO` in this case as `self.jetpackAccount` is nil.

File: WordPress/Classes/Models/Blog.m (hunk 3 above)
Comment (aerych) [inline WordPress/Classes/Models/Blog.m]: see above :)

Comment (aerych) [inline WordPress/Classes/Models/Blog.m]: Fwiw, I considered consolidating the two if blocks into something like 
`if ((self.dotComID || [self jetpackRESTSupported]) && [self jetpackSharingButtonsModleEnabled)` but I decided separating the checks was better for clarity of purpose.

File: WordPress/Classes/Models/Blog.m (hunk 2 above)
Comment (koke) [inline WordPress/Classes/Models/Blog.m]: I see, I forgot the first check was `HostedAtWPcom` and not just REST. I still find the logic a bit hard to follow. How about this?

```obj-c
    // Publicize is only supported via REST
    if (![self supports:BlogFeatureWPComRESTAPI]) {
        return NO;
    }

    if (self.isHostedAtWPcom) {
        // For WordPress.com YES unless it's disabled
        return ![[self getOptionValue:OptionsKeyPublicizeDisabled] boolValue];
    } else {
        // For Jetpack, check if the module is enabled
        return [self jetpackPublicizeModuleEnabled];
    }
```

Comment (aerych) [inline WordPress/Classes/Models/Blog.m]: Doesn't seem hard to follow to me, but then I wrote it. ;P
Sure. That works just as well so I'll make the change.

**Comment by aerych** [comment]
> Some other ideas

I really liked the `NSOperation` approach, but it would need the XMLRPC remotes to return NSOperations and that would require a pod change.  That's a bit more of a refactor than I'm looking to do here. 

Instead I've gone with the `dispatch_group` approach. Thanks for the tip!!  I love how light weight the change is.

Ready for another round @koke :D

File: WordPress/Classes/Models/Blog.m [hunk 5]
```diff
@@ -20,6 +20,11 @@
 static NSInteger const ImageSizeLargeHeight = 480;
 
 NSString * const PostFormatStandard = @"standard";
+NSString * const ActiveModulesKeyPublicize = @"publicize";
+NSString * const ActiveModulesKeySharingButtons = @"sharedaddy";
+NSString * const OptionsKeyActiveModules = @"active_modules";
+NSString * const OptionsKeyPublicizeDisabled = @"publicize-permanently-disabled";
```
Comment (koke) [inline WordPress/Classes/Models/Blog.m]: I think this should be `publicize_permanently_disabled`

**Comment by koke** [comment]
On a Jetpack site, with Publicize disabled and Sharing enabled, I get this:

simulator screen shot 02 feb 2016 08 42 29

With Publicise enabled and Sharing disabled there's no error. Although both sections are visible if one of the modules is disabled.

Then, possibly out of scope, Facebook connection fails. Or any other connection. But now it fixed itself and I can't reproduce :angry: 

simulator screen shot 02 feb 2016 08 45 53
simulator screen shot 02 feb 2016 08 45 56

**Comment by koke** [comment]
Just to leave a note, and related to #3761, I believe "Sharing" should not hide if Jetpack is not configured but, like Stats, it should be a prompt to configure Jetpack. It's kind of weird and unexpected that you have to go through stats to enable a different Jetpack feature.

File: WordPress/Classes/ViewRelated/Stats/JetpackSettingsViewController.m [hunk 6]
```diff
@@ -457,11 +457,15 @@ - (IBAction)saveAction:(id)sender
...
     void (^finishedBlock)() = ^() {
-        [self setAuthenticating:NO];
-        
-        if (self.completionBlock) {
-            self.completionBlock(YES);
-        }
+        // Ensure options are up to date after connecting Jetpack as there may
+        // now be new info.
+        BlogService *service = [[BlogService alloc] initWithManagedObjectContext:[[ContextManager sharedInstance] mainContext]];
+        [service syncBlog:self.blog completionHandler:^(NSArray *failures) {
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Stats/JetpackSettingsViewController.m]: This one is crashing because of the extra array parameter

**Comment by koke** [comment]
This one's weird. I logged in with a test wp.com account (kokejb7), then added a Jetpack site (wpt.koke.me) connected to a different account (kokeme), went to stats and connected Jetpack. Both Publicize and Sharing are enabled. The Jetpack site only has one connected account, and it's an admin. However I get this:

simulator screen shot 02 feb 2016 09 11 58

And when I try to connect:

simulator screen shot 02 feb 2016 09 11 45

If I break on the publicize sync failure and inspect the error:

```
Error Domain=com.wordpress.api Code=4 "This user is not authorized to access publicize connections on this site. [authorization_required]" UserInfo={WordPressComApiErrorCodeKey=authorization_required, NSLocalizedDescription=This user is not authorized to access publicize connections on this site. [authorization_required]}
```

**Comment by koke** [comment]
> Debugging it more, it seems it's using the wrong token (kokejb7)

Indeed, `SharingService.apiForRequest()` is always using the api from the default account, regardless of blog.

File: WordPress/Classes/ViewRelated/Stats/JetpackSettingsViewController.m (hunk 6 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Stats/JetpackSettingsViewController.m]: Oops. Missed that one when I refactored.

**Comment by aerych** [comment]
> On a Jetpack site, with Publicize disabled and Sharing enabled ...

Yep!  Much of the UI is still a work in progress. The sharing vc will be updated later to show/hide supported features in a subsequent PR. 

>  I believe "Sharing" should not hide if Jetpack is not configured but, like Stats, it should be a prompt to configure Jetpack

Agreed, and it should also include instructions on how to enable the relevant disabled Jetpack modules. Very much related to the note above. 

> Indeed, `SharingService.apiForRequest()` is always using the api from the default account, regardless of blog

DoH! Great catch right there. I've changed things so the blog's `restApi` is used. I also fixed some issues authenticating Jetpack connected blogs when making the oAuth hand shake. Maybe a bit out of scope for this PR but oh well.  

Ready for another round @koke :)

File: WordPress/Classes/Services/SharingService.swift [hunk 7]
```diff
@@ -451,25 +455,20 @@ public class SharingService : LocalCoreDataService
...
+    ///     - blog: The blog to use for the rest api.
+    ///
+    private func apiForBlog(blog: Blog) -> WordPressComApi {
+        if let api = blog.restApi() {
+            return api
         }
 
-        return api!
+        DDLogSwift.logError("SharingService is not using a real WordPress.com account.")
+        return WordPressComApi.anonymousApi()
```
Comment (koke) [inline WordPress/Classes/Services/SharingService.swift]: Did you drop the `assert` intentionally? Also `assert(false) == assertionFailure()` :)

Comment (aerych) [inline WordPress/Classes/Services/SharingService.swift]: Yep.  The assert doesn't make sense when we're intending to return the anonymous api. However, that `logError` should probably be a `logDebug` with a better message.

File: WordPress/Classes/Services/SharingService.swift [hunk 8]
```diff
@@ -451,25 +455,14 @@ public class SharingService : LocalCoreDataService
...
-            assert(false, error)
-        }
-
+    /// - Parameters:
+    ///     - blog: The blog to use for the rest api.
+    ///
+    private func apiForBlog(blog: Blog) -> WordPressComApi {
+        let api: WordPressComApi? =  blog.restApi()
+        assert(api != nil)
         return api!
```
Comment (koke) [inline WordPress/Classes/Services/SharingService.swift]: We talked about it on Slack, but leaving a note here, for future reference. I don't like the forced unwrap and style of this, but it's good for now. When we add nullability annotations to `Blog` (or convert it to swift :wink: ) we'll be able to avoid the cast to optional.

As for the force unwrap, I think it'd require passing the blog when initializing, so we can fail if blog.restApi is nil. But even then, the calling code would have to handle the possibility of a nil service.

Anyway, good to go, just wanted to write my thoughts somewhere :smile:

**Comment by koke** [comment]
:shipit:

**Comment by aerych** [comment]
Thanks Koke!!

File: WordPress/Classes/Services/SharingService.swift (hunk 8 above)
Comment (aerych) [inline WordPress/Classes/Services/SharingService.swift]: > we add nullability annotations to Blog [...] we'll be able to avoid the cast to optional.

I'll happily remove the optional cast when that happens!  For now I think the extra clarity the deliberate unwrapping provides is a good thing.  Thanks again :)
